docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,420 @@
+ */
+void oe_free_attestation_evidence(uint8_t* evidence_buffer);
+
+/**
+ * oe_get_attestation_collateral
+ *
+ * Generates the attestaton collateral for the given UUID attestation format.
+ *
+ * @param[in] evidence_format_uuid The UUID of the plugin.
+ * @param[out] collateral_buffer An output pointer that will be assigned the
+ * address of the collateral buffer.
+ * @param[out] collateral_buffer_size A pointer that points to the size of the
+ * collateral buffer.
+ * @retval OE_OK The function succeeded.
+ * @retval OE_NOT_FOUND The plugin does not exist.
+ */
+oe_result_t oe_get_attestation_collateral(
+    const uuid_t* evidence_format_uuid,
+    uint8_t** collateral_buffer,
+    size_t* collateral_buffer_size);
+
+/**
+ * oe_free_attestation_collateral
+ *
+ * Frees the attestation collateral.
+ *
+ * @param[in] collateral_buffer A pointer to the collateral buffer.
+ */
+void oe_free_attestation_collateral(uint8_t* collateral_buffer);
+
+/**
+ * oe_verify_attestation_evidence
+ *
+ * Verifies the attestation evidence and returns well known and custom claims.
+ *
+ * @param[in] evidence_buffer The evidence buffer.
+ * @param[in] evidence_buffer_size The size of evidence_buffer in bytes.
+ * @param[in] collateral_buffer The collateral buffer.
+ * @param[in] collateral_buffer_size The size of collateral_buffer in bytes.
+ * @param[out] claims The list of claims.
+ * @param[out] claims_size The size of claims.
+ * @retval OE_OK on success.
+ */
+oe_result_t oe_verify_attestation_evidence(
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    const uint8_t* collateral_buffer,
+    size_t collateral_buffer_size,
+    uint_t** claims,
+    size_t* claims_size);
+````
+
+User Experience
+---------------
+
+There are two types of users: the plugin writers and the plugin consumers.
+
+Plugin writers will implement their plugin according to the plugin API:
+
+```C
+/* Plugin implementation functions here. */
+
+/* Setting up the plugin struct. */
+oe_attestation_plugin_t my_plugin = {
+ /* Plugin UUID. */
+ .format_id = MY_PLUGIN_UUID,
+
+  /* Plugin functions. */
+ .on_register = my_plugin_on_register,
+  /* Rest of the plugin functions. */
+};
+```
+
+They can then compile their code in the standard way for building Open Enclave
+enclave and host applications.
+
+Plugin consumers will use the new "plugin aware" APIs like
+`oe_get_attestation_evidence`. Plugin consumers will use the APIs like this:
+
+```C
+/* Register plugin. */
+oe_register_plugin(my_plugin, my_config_data, my_config_data_size);
+
+/* Get evidence. */
+oe_get_attestation_evidence(
+    MY_PLUGIN_UUID,
+    my_custom_claims,
+    my_custom_claims_size,
+    &evidence,
+    &evidence_size);
+
+/* Get collateral. */
+oe_get_attestation_collateral(
+    MY_PLUGIN_UUID,
+    &collateral,
+    &collateral_size);
+
+/* Verify evidence. Can check the clams if desired. */
+oe_verify_attestaton_evidence(
+    evidence,
```

Comment:
What happened to the MY_PLUGIN_UUID here?

Reply:
It's part of the header, so it's unneeded. I'll add a section about the header.